src/provider/meta.ts

```diff
@@ -0,0 +1,63 @@
+import type { ExternalProvider, JsonRpcProvider, Web3Provider } from '@ethersproject/providers'
+import WalletConnectProvider from '@walletconnect/ethereum-provider'
+
+function isWeb3Provider(provider: JsonRpcProvider): provider is Web3Provider {
+  return 'provider' in provider
+}
+
+function isWalletConnectProvider(provider: ExternalProvider): provider is WalletConnectProvider {
+  return (provider as WalletConnectProvider).isWalletConnect
+}
+
+export enum WalletType {
+  WALLET_CONNECT = 'WalletConnect',
+  INJECTED = 'Injected',
+  UNKNOWN = 'Unknown',
+}
+
+export interface WalletMeta extends Partial<NonNullable<WalletConnectProvider['connector']['peerMeta']>> {
+  type: WalletType
+}
+
+function getWalletConnectMeta(provider: WalletConnectProvider): WalletMeta {
+  return {
+    type: WalletType.WALLET_CONNECT,
+    ...provider.connector.peerMeta,
+  }
+}
+
+function getInjectedMeta(provider: ExternalProvider & Record<string, unknown>): WalletMeta {
+  const properties = Object.getOwnPropertyNames(provider)
+  const names =
+    properties
+      .filter((name) => name.match(/^is.*$/) && (provider as Record<string, unknown>)[name] === true)
+      .map((name) => name.slice(2)) ?? []
+
+  // Add qrUrl to identify QR-initiated connections (eg for Coinbase Wallet via mobile QR).
+  if (properties.includes('qrUrl') && provider['qrUrl']) {
+    names.push('qrUrl')
+  }
+
+  // Sort MetaMask last, so that wallets spoofing MetaMask list themselves first.
+  names.sort((a, b) => (a === 'MetaMask' ? 1 : b === 'MetaMask' ? -1 : 0))
+
+  return {
+    type: WalletType.INJECTED,
+    name: names.join(' '),
+    // TODO(WEB-2914): Populate description, url, and icons for known wallets.
+  }
+}
+
+export function getWalletMeta(provider: JsonRpcProvider): WalletMeta {
+  if (!isWeb3Provider(provider)) return { type: WalletType.UNKNOWN }
```

Comment:
If we can't process meta for a given wallet, maybe we should throw an error instead?

Reply:
I don't think we should throw an error, because that would need to be caught. I'd rather have an `UNKNOWN` type so that it doesn't trip up uses where the unknown is not handled.

Reply:
I don't think we should throw, because that would require users to catch for a utility method that just gets metadata.
If we can't process data, then it is not an external provider, and is not a wallet. I think we should instead return `undefined`.